core/src/main/java/org/mule/runtime/core/internal/construct/DefaultFlowBuilder.java

```diff
@@ -0,0 +1,297 @@
+   * @return same builder instance.
+   */
+  public Builder messageSource(MessageSource messageSource) {
+    checkImmutable();
+    checkArgument(messageSource != null, "messageSource cannot be null");
+    this.messageSource = messageSource;
+
+    return this;
+  }
+
+  /**
+   * Configures the message processors to execute as part of flow.
+   *
+   * @param messageProcessors message processors to execute. Non null.
+   * @return same builder instance.
+   */
+  public Builder messageProcessors(List<Processor> messageProcessors) {
+    checkImmutable();
+    checkArgument(messageProcessors != null, "messageProcessors cannot be null");
+    this.messageProcessors = messageProcessors;
+
+    return this;
+  }
+
+  /**
+   * Configures the exception listener to manage exceptions thrown on the flow execution.
+   *
+   * @param exceptionListener exception listener to use on the flow. Non null.
+   * @return same builder instance
+   */
+  public Builder messagingExceptionHandler(MessagingExceptionHandler exceptionListener) {
+    checkImmutable();
+    checkArgument(exceptionListener != null, "exceptionListener cannot be null");
+    this.exceptionListener = exceptionListener;
+    return this;
+  }
+
+  /**
+   * Configures the factory used to create processing strategies on the created flow.
+   *
+   * @param processingStrategyFactory factory to create processing strategies. Non null.
+   * @return same builder instance.
+   */
+  public Builder processingStrategyFactory(ProcessingStrategyFactory processingStrategyFactory) {
+    checkImmutable();
+    checkArgument(processingStrategyFactory != null, "processingStrategyFactory cannot be null");
+    this.processingStrategyFactory = processingStrategyFactory;
+    return this;
+  }
+
+  /**
+   * Builds a flow with the provided configuration.
+   *
+   * @return a new flow instance.
+   */
+  public Flow build() {
+    checkImmutable();
+
+    flow = new DefaultFlow(name, muleContext);
+    if (messageSource != null) {
+      flow.setMessageSource(messageSource);
+    }
+
+    if (messageProcessors != null) {
+      flow.setMessageProcessors(messageProcessors);
+    }
+
+    if (exceptionListener != null) {
+      flow.setExceptionListener(exceptionListener);
+    }
+
+    if (processingStrategyFactory != null) {
+      flow.setProcessingStrategyFactory(processingStrategyFactory);
+    }
+
+    if (muleContext.getRegistry() != null) {
+      try {
+        muleContext.getRegistry().registerFlowConstruct(flow);
+      } catch (MuleException e) {
+        throw new IllegalStateException("Cannot register flow in Mule context", e);
+      }
+    }
+
+    return flow;
+  }
+
+  protected final void checkImmutable() {
+    if (flow != null) {
+      throw new IllegalStateException("Cannot change attributes once the flow was built");
+    }
+  }
+
+  /**
+   * Default implementation of {@link Flow}
+   */
+  public static class DefaultFlow extends AbstractPipeline implements Flow {
```

Comment:
Isn't it best for tests to extend Flow to create a builder and create they own implementation of flow that way?  This would mean this class could be protected.